src/main/java/com/example/java/maven/game2048/Board.java

```diff
@@ -18,18 +23,267 @@ public Board(int numberOfRows, int numberOfColumns) {
+                    return true;
+
+                }
+            }
+        }
+        return false;
+    }
+
+    public boolean checkIfSimilarNeighboringFields() {
+        for (Field[] row : this.fields) {
+            for (int index = 0; index < row.length - 1; index++) {
+                if (row[index].getValue() == row[index + 1].getValue()) {
+                    return true;
+                }
+            }
+        }
+        for (Field[] column : getColumns()) {
+            for (int index = 0; index < column.length - 1; index++) {
+                if (column[index].getValue() == column[index + 1].getValue()) {
+                    return true;
+                }
+            }
+        }
+        return false;
+    }
+
+    public boolean checkIfGameOver() {
+        int zeroValuesOnBoard = getFieldsWithZeroValue().size();
+        if (zeroValuesOnBoard > 0) {
+            return false;
+        }
+        if (checkIfSimilarNeighboringFields()) {
+            return false;
+        }
+        System.out.println("YOU LOST!");
+        return true;
+    }
+
+    private Field[][] getCopyOfFields(Board toCopy) {
+        Field[][] copy = new Field[numberOfRows][numberOfColumns];
+        for (int rowNumber = 0; rowNumber < numberOfColumns; rowNumber++) {
+            for (int columnNumber = 0; columnNumber < numberOfRows; columnNumber++) {
+                copy[rowNumber][columnNumber] = new Field(toCopy.fields[rowNumber][columnNumber]);
+            }
+        }
+        return copy;
+    }
+
+    private Field[] getColumn(int columnNumber) {
+        Field[] column = new Field[numberOfRows];
+        for (int rowNumber = 0; rowNumber < numberOfRows; rowNumber++) {
+            column[rowNumber] = fields[rowNumber][columnNumber];
+        }
+        return column;
+    }
+
+    private Field[][] getColumns() {
+        Field[][] columns = new Field[numberOfRows][numberOfColumns];
+        for (int columnNumber = 0; columnNumber < numberOfColumns; columnNumber++) {
+            columns[columnNumber] = getColumn(columnNumber);
+        }
+        return columns;
     }
+
+    private List<Field> getFieldsWithZeroValue() {
+        List<Field> fieldsWithZeroValue = new ArrayList<>();
+        for (Field[] row : this.fields) {
+            for (Field field : row) {
+                if (field.getValue() == 0) {
+                    fieldsWithZeroValue.add(field);
+                }
+            }
+        }
+        return fieldsWithZeroValue;
+    }
+
+    private Field getRandomFieldWithZeroValue() {
+        Random random = new Random();
+        List<Field> fieldsWithZeroValues = getFieldsWithZeroValue();
+        return fieldsWithZeroValues.get(random.nextInt(fieldsWithZeroValues.size()));
+    }
+
+    private boolean notSortedLeftOrUp(Field[] rowOrColumn) {
+        for (int index = 0; index < rowOrColumn.length - 1; index++) {
+            if (rowOrColumn[index].getValue() == 0 && rowOrColumn[index + 1].getValue() > 0) {
```

Comment:
minor - introduce variables which simplify code: `currentFieldValue` and `nextFieldValue` - check this class and apply where possible

Reply:
Done